File: src/main/java/org/hisp/dhis/rules/variables/ProgramRuleCustomVariable.java
package org.hisp.dhis.rules.variables;

import org.hisp.dhis.antlr.ParserExceptionWithoutContext;
import org.hisp.dhis.rules.RuleExpression;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;

import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;

public class ProgramRuleCustomVariable
    extends ScalarFunctionToEvaluate
{
    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        RuleVariableValue variableValue = visitor.getValueMap().get( ctx.programRuleVariableName().getText() );

        if ( variableValue == null )
        {
            throw new ParserExceptionWithoutContext(
                "Variable " + ctx.programRuleVariableName().getText() + " not present" );
        }

        return variableValue.value() == null ?
                variableValue.type().defaultValue() : getValueCastedByType( variableValue );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        String variable = null;

        if ( ctx.programRuleVariableName() != null )
        {
            variable = ctx.programRuleVariableName().getText();
        }
        else if ( ctx.programRuleStringVariableName() != null )
        {
            variable = ctx.programRuleStringVariableName().getText().replace("\'","");
        }

        if ( visitor.getItemStore().containsKey( variable ) )
        {
            visitor.getItemDescriptions().put( ctx.getText(), visitor.getItemStore().get( variable ).getDisplayName() );

            return visitor.getItemStore().get( variable ).getValueType().getValue();
        }

        throw new ParserExceptionWithoutContext(
            "Variable " + variable + " does not exist" );
    }
}

File: src/jvmMain/kotlin/org/dhis2/ruleengine/variables/ProgramRuleCustomVariable.kt
package org.dhis2.ruleengine.variables

import org.dhis2.ruleengine.parser.expression.CommonExpressionVisitor
import org.dhis2.ruleengine.parser.expression.function.ScalarFunctionToEvaluate
import org.hisp.dhis.antlr.ParserExceptionWithoutContext
import org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext

class ProgramRuleCustomVariable : ScalarFunctionToEvaluate() {
    override fun evaluate(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variableValue = visitor.getValueMap()[ctx.programRuleVariableName().text]
            ?: throw ParserExceptionWithoutContext(
                "Variable " + ctx.programRuleVariableName().text + " not present"
            )
        return variableValue.value ?: variableValue.ruleValueType.defaultValue
    }

    override fun getDescription(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        var variable: String? = null
        if (ctx.programRuleVariableName() != null) {
            variable = ctx.programRuleVariableName().text
        } else if (ctx.programRuleStringVariableName() != null) {
            variable = ctx.programRuleStringVariableName().text.replace("\'", "")
        }
        if (visitor.getItemStore().containsKey(variable)) {
            visitor.itemDescriptions[ctx.text] = visitor.getItemStore()[variable]!!.displayName
            return visitor.getItemStore()[variable]!!.valueType.value
        }
        throw ParserExceptionWithoutContext(
            "Variable $variable does not exist"
        )
    }
}